lib/uknetdev/netdev.c

```diff
@@ -274,44 +301,74 @@ void uk_netdev_info_get(struct uk_netdev *dev,
 				      dev_info->max_tx_queues);
 }
 
-static const void *_netdev_einfo_get(struct uk_netdev *dev,
+const char *uk_netdev_einfo_get(struct uk_netdev *dev,
 				enum uk_netdev_einfo_type einfo)
 {
-	switch (einfo) {
-	case UK_NETDEV_IPV4_ADDR_STR:
-		if (dev->_einfo->ipv4_addr)
-			uk_pr_debug("ip: %s\n", dev->_einfo->ipv4_addr);
-		return dev->_einfo->ipv4_addr;
-	case UK_NETDEV_IPV4_MASK_STR:
-		if (dev->_einfo->ipv4_net_mask)
-			uk_pr_debug("netmask: %s\n", dev->_einfo->ipv4_net_mask);
-		return dev->_einfo->ipv4_net_mask;
-	case UK_NETDEV_IPV4_GW_STR:
-		if (dev->_einfo->ipv4_gw_addr)
-			uk_pr_debug("gateway: %s\n", dev->_einfo->ipv4_gw_addr);
-		return dev->_einfo->ipv4_gw_addr;
-	case UK_NETDEV_IPV4_DNS0_STR:
-		if (dev->_einfo->ipv4_dns0_addr)
-			uk_pr_debug("dns: %s\n", dev->_einfo->ipv4_dns0_addr);
-		return dev->_einfo->ipv4_dns0_addr;
-	default:
-		uk_pr_warn("Option %d not yet supported\n", einfo);
-	}
-	return NULL;
-}
+	const char *ret = NULL;
 
-const void *uk_netdev_einfo_get(struct uk_netdev *dev,
-				enum uk_netdev_einfo_type einfo)
-{
 	UK_ASSERT(dev);
 	UK_ASSERT(dev->ops);
 	UK_ASSERT(dev->_data->state >= UK_NETDEV_UNCONFIGURED);
 
-	if (dev->_einfo)
-		return _netdev_einfo_get(dev, einfo);
-	else if (dev->ops->einfo_get)
-		return dev->ops->einfo_get(dev, einfo);
-	return NULL;
+#if CONFIG_LIBUKNETDEV_EINFO_LIBPARAM
+	if (dev->_einfo) {
+		switch (einfo) {
+		case UK_NETDEV_IPV4_ADDR:
+			if (dev->_einfo->ip4.cidr ||
+			    (dev->ops->einfo_get &&
+			     dev->ops->einfo_get(dev, UK_NETDEV_IPV4_CIDR)))
+				return NULL; /* CIDR (overwrite) exists */
+			break;
+		case UK_NETDEV_IPV4_MASK:
+			if (dev->_einfo->ip4.cidr ||
+			    (dev->ops->einfo_get &&
+			     dev->ops->einfo_get(dev, UK_NETDEV_IPV4_CIDR)))
+				return NULL; /* CIDR (overwrite) exists */
+			break;
+		case UK_NETDEV_IPV4_CIDR:
+			if (dev->_einfo->ip4.cidr)
+				return dev->_einfo->ip4.cidr;
+			break;
+		case UK_NETDEV_IPV4_GW:
+			if (dev->_einfo->ip4.gw)
+				return dev->_einfo->ip4.gw;
+			break;
+		case UK_NETDEV_IPV4_DNS0:
+			if (dev->_einfo->ip4.dns0)
+				return dev->_einfo->ip4.dns0;
+			break;
+		case UK_NETDEV_IPV4_DNS1:
+			if (dev->_einfo->ip4.dns1)
+				return dev->_einfo->ip4.dns1;
+			break;
+		case UK_NETDEV_IPV4_HOSTNAME:
+			if (dev->_einfo->ip4.hostname)
+				return dev->_einfo->ip4.hostname;
+			break;
+		case UK_NETDEV_IPV4_DOMAIN:
+			if (dev->_einfo->ip4.domain)
+				return dev->_einfo->ip4.domain;
+			break;
+		default:
+		}
+	}
+#endif /* CONFIG_LIBUKNETDEV_EINFO_LIBPARAM */
+
+	if (!ret && dev->ops->einfo_get) {
```

Comment:
Isn't `ret` always `NULL` here?

Reply:
Actually, true. Probably a left-over. I'll remove it.

Reply:
Hey, in the new version I got rid of `ret` variable completely. Please, double-check. I also noticed that `NULL` would be always returned in case of `UK_NETDEV_IPV4_ADDR` and `UK_NETDEV_IPV4_MASK`. However, there should be a value returned by `dev->ops->einfo_get()` if there is no CIDR address with the device's einfo or the user provided kernel parameter.